Add "debug" module and many debug() calls to help

emhagman opened the pull request:
I believe this can help us greatly when trying to find new bugs. People will be able to enable debugging via `env DEBUG="agenda:*" node index.js` or `DEBUG="agenda:*"` and paste the output into GitHub to help us determine what is going on.

The `debug` module is very well known and supported and I believe it is a reasonable dependency to pull in. I've gone ahead and added many `debug()` statements that I think will be useful.

Please let me know if there is anything I can do to help get this merged!

emhagman commented:
I believe the `codeclimate` findings are a little silly since the "duplicates" are either debug statements or function declarations that already exist that, since they have been cleaned up, now trip the warning.

Let me know if I should change anything!

simison commented:
Great! A few questions:

- A very generic `DEBUG` env var sounds a bit funny to me but you say it's a standard'ish thing? An alternative could be something like `agenda.set('debug', true)` or `agenda.debug(true)` but I'm good either way.

- I'm not very familiar with Yarn yet, is it normal/ok to upgrade `yarn.lock` with PRs or should it be done separately by maintainers?

- I was about to ask if you could mention debugging at README.md as well, but I just realised we don't have a chapter there for development at all. :-)

OmgImAlexis commented:
I'll review this when I get a chance. I'd like to change coveralls to 2% or maybe 3% at least while where going through the cleanup stage as most PRs like this are going to end up with coveralls failing even though it's not really a fail per se.

OmgImAlexis commented:
@simison Using the DEBUG env like @emhagman has stated with this package is standard and what a lot of other large packages use.

The yarn.lock should be added by a maintainer to start with but adding it in a PR isn't that bad, I'll likely add one myself in a bit and then rebase this on top of it so we only end up adding debug to `yarn.lock`.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -76,6 +90,7 @@ Agenda.prototype.database = function(url, collection, options, cb) {
   MongoClient.connect(url, options, function( error, db ){
     if (error) {
       if (cb) {
+        debug('connected to %')
```
Was this meant to have a var added as the second param?

simison replied:
BTW just in case someone thinks of adding `url` to that debug statement:
I'd be hesitant to log `url` here since it might contain user+pass.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -287,6 +320,7 @@ Agenda.prototype.define = function(name, options, processor) {
  * @returns {*} Job or jobs created
  */
 Agenda.prototype.every = function(interval, names, data, options, cb) {
+
```
Remove unneeded line.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -439,8 +488,14 @@ Agenda.prototype.now = function(name, data, cb) {
+      if (error) {
+        debug('error trying to delete jobs from MongoDB');
+      } else {
+        debug('job cancellation was successful!');
```
Remove `!` to stay consistent with other debug statements.

simison replied:
This could be just "_job cancelled_"?

I kinda like idea adding job name or id there but not sure if that would actually be useful or just add noise.

OmgImAlexis replied:
Since the name should be above I think it'll be fine.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -468,25 +525,30 @@ Agenda.prototype.saveJob = function(job, cb) {
 
     // Job type set to 'single' so...
     // NOTE: Again, not sure about difference between 'single' here and 'once' in job.js
+    debug('job with type of "single" found!');
```
Remove `!` to stay consistent with other debug statements.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -468,25 +525,30 @@ Agenda.prototype.saveJob = function(job, cb) {
     // If the nextRunAt time is older than the current time, "protect" that property, meaning, don't change
     // a scheduled job's next run time!
     if (props.nextRunAt && props.nextRunAt <= now) {
+      debug('job has a scheduled nextRunAt time, protecting that field from upsert!');
```
Remove `!` to stay consistent with other debug statements.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -559,6 +627,7 @@ Agenda.prototype.saveJob = function(job, cb) {
 
         // If the current job would have been processed in an older scan, process the job immediately
         if (job.attrs.nextRunAt && job.attrs.nextRunAt < self._nextScanAt) {
+          debug('[%s:%s] job would have ran by nextScanAt, processing the job immediately!', job.attrs.name, res._id);
```
Remove `!` to stay consistent with other debug statements.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -579,8 +648,11 @@ Agenda.prototype.saveJob = function(job, cb) {
     this._processInterval = setInterval(processJobs.bind(this), this._processEvery);
     process.nextTick(processJobs.bind(this));
+  } else {
+    debug('Agenda.start was already called, ignoring!');
```
Remove `!` to stay consistent with other debug statements.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -649,8 +724,15 @@ Agenda.prototype._findAndLockNextJob = function(jobName, definition, cb) {
+          debug('error occurred when running query to find and lock job');
+        } else {
+          // debug('no jobs to unlock found!');
+          // NOTE: This can get called a lot, do we want to have it here?
```
I'd say skip this for now, we can always add it back if we need.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -694,6 +778,7 @@ function processJobs(extraJob) {
   // Make sure an interval has actually been set
   // Prevents race condition with 'Agenda.stop' and already scheduled run
   if (!this._processInterval) {
+    debug('no _processInterval set when calling processJobs, most likely Agenda.stop() called');
```
The last bit needs to be reworded.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -705,13 +790,15 @@ function processJobs(extraJob) {
   } else if (definitions[extraJob.attrs.name]) {
 
     // Add the job to list of jobs to lock and then lock it immediately!
+    debug('extra job was passed in, immediate lock and execute to occur!');
```
This may need to be reworded.

Remove `!` to stay consistent with other debug statements.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -733,6 +820,7 @@ function processJobs(extraJob) {
     if (jobDefinition.lockLimit && jobDefinition.lockLimit <= jobDefinition.locked) {
       shouldLock = false;
     }
+    debug('checking lock status for [%s]... should lock? %s', name, shouldLock);
```
May need to be reworded.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -857,17 +949,22 @@ function processJobs(extraJob) {
     self._nextScanAt = new Date(now.valueOf() + self._processEvery);
 
+    // NOTE: This may print a lot based on _processInterval
+    debug('_nextScanAt set to [%s]', self._nextScanAt.toISOString());
```
Let's leave this in for now and see how it does. I'm guessing this would only really be a problem with short process intervals?

lushc replied:
The default process interval is 5 seconds so it might be quite spammy, but yeah we won't be able to tell until someone gives us some output to look at.

OmgImAlexis replied:
@niftylettuce this is the only issue I have with this. @lushc are you okay with merging this and if need be removing this later on if we find it's too spammy? Since it'll only show when debug is enabled we should be okay with it.

niftylettuce replied:
I think this would be too spammy IMO.

OmgImAlexis replied:
@emhagman would you mind removing this for now and if we need to we can always add it back.

emhagman replied:
@OmgImAlexis Sure thing!

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -857,17 +949,22 @@ function processJobs(extraJob) {
     self._findAndLockNextJob(name, definitions[name], function(err, job) {
 
       if (err) {
+        debug('error trying to find a job to lock in queue filling process');
```
May need to be reworded.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -877,6 +974,7 @@ function processJobs(extraJob) {
       // 3. Queue the job to actually be run now that it is locked
       // 4. Recursively run this same method we are in to check for more available jobs of same type!
       if (job) {
+        debug('found a job to lock during queue filling!');
```
May need to be reworded.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -910,13 +1008,17 @@ function processJobs(extraJob) {
     // If the 'nextRunAt' time is older than the current time, run the job
     // Otherwise, setTimeout that gets called at the time of 'nextRunAt'
     if (job.attrs.nextRunAt < now) {
+      debug('[%s:%s] nextRunAt is in the past, run the job immediately!', job.attrs.name, job.attrs._id);
```
Remove `!` to stay consistent with other debug statements.

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -910,13 +1008,17 @@ function processJobs(extraJob) {
     } else {
-      setTimeout(runOrRetry, job.attrs.nextRunAt - now);
+      var runIn = job.attrs.nextRunAt - now;
+      debug('[%s:%s] nextRunAt is in the future, call setTimeout(%d)', job.attrs.name, job.attrs._id, runIn);
```
Replace "call" with "calling".

OmgImAlexis commented on lib/agenda.js:
```diff
@@ -954,6 +1058,7 @@ function processJobs(extraJob) {
         } else {
 
           // Run the job immediately by putting it on the top of the queue
+          debug('[%s:%s] concurrency preventing immediate run, push job to top of queue', job.attrs.name, job.attrs._id);
```
push -> pushing

OmgImAlexis commented on package.json:
```diff
@@ -28,6 +28,7 @@
     "url": "https://github.com/agenda/agenda/issues"
   },
   "dependencies": {
+    "debug": "^2.6.8",
```
Should we not add this to devDeps?

emhagman replied:
I don't believe so. You want people to be able to post logs if they have problems and they won't be able to turn it on on their server if it is in devDeps.

See in `express`: https://github.com/expressjs/express/blob/master/package.json#L36

OmgImAlexis replied:
Actually since we require it in the file it'll need to be in deps otherwise we'll get issues it being missing. Ignore that.

emhagman commented:
@OmgImAlexis I will remove the `yarn.lock` for now, didn't mean to slip that in! I'll get back to you with more changes

emhagman commented:
@OmgImAlexis I've updated the language and removed `yarn.lock`. See my comment above about the `devDependencies` issue. Let me know if there is anything else!

OmgImAlexis commented:
Can you please rebase since there's a conflicting issue with `lib/job.js` and you'll need to copy the yarn.lock from master to fix the conflict with that.

@simison @lushc for the moment we may just have to override merges because codeclimate is being a pain.

emhagman commented:
@OmgImAlexis  Apologies, my git skills are not the best but I believe everything is all set now for the merge.

OmgImAlexis commented:
I'm too tired to merge this without errors so I'll have a look first thing tomorrow morning unless one of the other devs gets to it first.

simison commented:
@OmgImAlexis @emhagman is this ready for merge? If we can get this in I think we can close [0.10.0](https://github.com/agenda/agenda/milestone/1) and get it published before "big breaking" [1.0.0](https://github.com/agenda/agenda/milestone/4) that should probably come next.

simison commented:
@emhagman wanna PR an update to README.md and History.md about this as well? Just so that it's not a hidden feature. Could even be in this PR, either way.

simison commented:
@OmgImAlexis when merging, remember to choose "squash and merge" so that we'll get clean commit log. :-)

emhagman commented:
@simison Sure thing, I can get to that at some point today!

emhagman commented:
@simison @OmgImAlexis Just added a "Debugging Issues" section to this PR. Let me know if any copy should change but other than that should be good to go!

OmgImAlexis commented:
Waiting on @lushc's reply for the spammy comment above and then I'm good to merge.

simison commented:
LGTM

@emhagman thanks, docs look great! I think I'll rearrange readme quite a bit in a separate PR later on and create stuff like [support](https://github.com/blog/2400-support-file-support) and contributing files.

I leave considering what Codeclimate & coverage errors mean to others.

emhagman commented:
@simison Thanks! As far as I can tell, since `createJob` now has the exact same function signature, it is made because of code duplication. I did not want to factor out this function in this PR because it seemed unrelated.

emhagman commented:
@OmgImAlexis I removed the spammy log. Is this good to go?

OmgImAlexis commented:
@emhagman yes sorry I've been MiA.

Since this adds a load of lines that can't really be tested it'll lower code coverage without us really being able to fix it so I'm going to merge even without all checks passing.
